fix: properly handle newlines with input when using the spinner

mbtools opened the pull request:
TL;DR

I fixed the spinner, yay! If you like it, [buy me a ☕ or 🍕 ](https://github.com/sponsors/mbtools). 

### Background

In 10.7.0, a [spinner](https://github.com/npm/cli/pull/7432) was introduced to the npm cli, a small progress indicator showing that npm is still busy and not dead.

I'm *not* a JavaScript developer, but I love a good challenge: Debugging and finding a fix for annoying bugs that no one had time to or wanted to fix is as good as sx (well, almost). 

### Problem

With the spinner came a regression that disrupted the user experience when npm asked for input during login or adding a user: The spinner would show up after the second prompt and drive users nuts.

For example, the username prompt for `npm adduser` would be ok but subsequent prompts for password and email showed the spinner although users were asked to enter something. Very confusing.

![Cursor_2354](https://github.com/user-attachments/assets/eddc7e70-f207-4e51-b0d7-0380dc9922f1)

### Root Cause

The was a race condition between turning off the spinner and showing a new prompt.

I tried setting timers, waiting for the next tick, and a flag indicating if there's an active prompt. None worked.

### Solution

This PR improves the cli prompt experience by ensuring the progress spinner does not appear between consecutive user prompts (such as username and password).

- Fixes race condition in input handler
- Avoids spinner interfering with input prompts
- Properly adds newlines with silent prompts or when input is cancelled by user

### Demo

Here's a demo of what it looks after this PR 😃 (uses [Verdaccio](https://verdaccio.org/) in case you wonder)

![Cursor_2345](https://github.com/user-attachments/assets/3f0ab224-ddf8-41ed-88ba-fc7201919544)

### References

Ref https://github.com/npm/cli/pull/7432
Closes https://github.com/npm/cli/issues/7425
Closes https://github.com/npm/cli/issues/7583

Enjoy, [Marc](https://github.com/sponsors/mbtools)

mbtools committed:
fix: prevent spinner from appearing between consecutive prompts

mbtools committed:
Update prompt, stop cursor blinking

mbtools committed:
Add exithandler

mbtools committed:
Fix display test

mbtools committed:
Add cursor hidden flag and global exithandler

mbtools committed:
Move cursor blink logic to separate PR

mbtools committed:
Merge branch 'latest' into spinner

mbtools committed:
Adjust snapshot

wraithgar commented on lib/utils/display.js:
```diff
@@ -357,10 +362,7 @@ class Display {
-          .catch(rej)
-          // Any call to procLog.input.read will render a prompt to the user, so we always
-          // add a single newline of output to stdout to move the cursor to the next line
-          .finally(() => output.standard('')))
```
We still need this for all other input prompts, including the initial password prompt.

mbtools replied:
do we? for example, with `npm init`, when I hit `enter`, it jumps to a newline (win, powershell).

wraithgar commented on lib/utils/read-user-info.js:
```diff
@@ -24,6 +24,9 @@ function readOTP (msg = otpPrompt, otp, isRetry) {
   if (isRetry && password) {
+    // because the prompt is silent and we need to move the cursor
+    // to the next line so the next output is on a new line.
+    process.stdout.write('\n')
```
We never write to stdout directly in npm.  All input and output goes through their respective functions in `proc-log` which are event-driven and intercepted by npm in its display layer.

There already was an effective newline in `lib/utils/display.js` that was covering this use case whenever we intercepted an "input" event from proc-log.  This is needed because the newline that a user enters is not acted on in and of itself, it's consumed by readline as part of the end of input.  If that isn't working I don't understand how this will.  You can also see in snapshots where that newline is now missing from all other input prompts.


```
const read = (...args) => input.read(() => _read(...args))
```

`input.read` is what emits `input.KEYS.read` and is then intercepted by npm, writing that newline via `output.standard('')`.


I think you may have been on the right track with the counter for active inputs.  I don't understand right now how we could have two active inputs, it may be a race condition.  However ALL of the functionality in there (not just the progress spinner) would need to be in the consideration for pausing until all input prompts are finished.

mbtools replied:
for me (win), only the silent password prompt consumes the newline at end of input (which is common for making sure the password won't be visible). the other prompts, work well without any of this. i understand that process.stdout isn't proper but didn't know how to do it otherwise (just for the pwd prompt). maybe there's some platform difference here?

i didn't see how two prompts could be active but the counter says otherwise ;-).

mbtools replied:
i tested with cmd, pwsh, and bash and it works well as is. each input is on it's own line. I don't get the point of the additional newline.

wraithgar replied:
This is a place where we are definitely going to have to lean on the "Chesterton's Fence" principle here and not remove something unless we can explain why it's there.  This would require testing in linux, osx and windows in multiple versions of each, and likely a deeper dive into the history of this code to see if it's not a legacy from a previous input method.

Please keep that output line in the display layer.


Going back to the process.stdout itself, that is going to have to be an emitted newline.  It should also be something we can see if we can work into the display layer itself.  This is not the only place where we will potentially be inputting silently, and if we did that in the future we would almost certainly forget to add this new line.  Perhaps we can add a flag on the input meta to let the display layer know it was a silent input?

mbtools replied:
The `finally` was added as part of https://github.com/npm/cli/pull/7432 which introduced the spinner. I think there's a good chance that it was put into the `inputHandler` because of the side-effect that the spinner shows up on the second and subsequent prompts. Now that this PR makes sure that the spinner no longer gets in the way, the `finally` is not necessary in general but only for silent prompts. maybe @lukekarrys remembers?

I will work on moving the newline for silent prompts to the display layer.

wraithgar replied:
Thanks for doing that history search.

lukekarrys replied:
iirc, the newline was written to move the cursor so no output would be overwritten if the user does a `ctrl+c` during one of the prompts.

Here's an example from `npm@11.4.1` and this PR where I `ctrl+c` during the `Username` prompt:

```
╰❯ npm adduser --auth-type=legacy
npm notice Log in on https://registry.npmjs.org/
Username:
npm error canceled
npm error A complete log of this run can be found in: /Users/lukekarrys/.cache/npm/_logs/2025-05-27T22_52_17_565Z-debug-0.log
```

```
╰❯ node . adduser --auth-type=legacy
npm notice Log in on https://registry.npmjs.org/
npm error canceled
npm error A complete log of this run can be found in: /Users/lukekarrys/.cache/npm/_logs/2025-05-27T22_52_23_380Z-debug-0.log
```

You can see in the second example from this PR that the `Username:` line get overwritten by the first line of the error. I believe this might be the only reason for the `.finally()` block. A small thing in comparison to the usability of multi-prompt commands I think, but maybe this context is helpful in deciding if it should stay or go.

wraithgar replied:
THANK YOU. We need to add the ctrl-c consideration to the comments so we know why it's there.  I think it needs to stay.

The newline on password prompts can be a separate concern still and `mbtools` is working on move that into the display layer.

wraithgar commented on lib/utils/auth.js:
```diff
@@ -51,7 +51,7 @@ const adduser = async (npm, { creds, ...opts }) => {
     const username = await read.username('Username:', creds.username)
     const password = await read.password('Password:', creds.password)
-    const email = await read.email('Email: (this IS public) ', creds.email)
+    const email = await read.email('Email (this will be public):', creds.email)
```
this is an unrelated `docs:` change that we don't want to hide in this PR under the `fix:`

mbtools committed:
Revert email prompt change

mbtools committed:
Add handling of silent prompts to display; add tests

mbtools committed:
Simplify

mbtools committed:
Switch to sequential input management

mbtools committed:
Revert token test changes

mbtools commented:
My AI friends and I spend a long time to figure this out, but we found a solution that is straight forward.

### Details

Here are the input events as they are processed currently, for example with `npm adduser`:

```
input.KEYS.read
input.KEYS.start
Username: me
input.KEYS.read.finally
input.KEYS.read
input.KEYS.start
Password: input.KEYS.end
input.KEYS.read 
input.KEYS.start
Email (this will be public): input.KEYS.read.finally
input.KEYS.end
me@me.com
input.KEYS.read.finally
input.KEYS.end
```

When `input.KEYS.read` is processed, it calls `input.start()` which immediately emits `input.KEYS.start`. However, the corresponding `input.KEYS.end` is only emitted when the promise returned by `input.start()` settles (due to the `.finally()` mechanism in `proc-log`  `input.start` function).

This creates a race condition where:

- The next prompt's `input.KEYS.read` and `input.KEYS.start` are emitted synchronously
- The previous prompt's `input.KEYS.end` is emitted asynchronously when the promise chain completes

This explains why we see the events out of order:

- `input.KEYS.read` (username)
- `input.KEYS.start` (username)
- `input.KEYS.read.finally` (username)
- `input.KEYS.read` (password) - next prompt starts
- `input.KEYS.start` (password) - next prompt starts
- `input.KEYS.end` (username) - previous prompt finally ends

The issue is that the `input.KEYS.end` event is tied to the promise resolution timing rather than being emitted immediately when the prompt completes, causing the overlapping event sequence.

### Solution

We switched to sequential input management because:

- *Minimal changes*: Only requires modifying the display handler
- *Clear event ordering*: Ensures start/end events are properly sequenced
- *Backward compatible*: Doesn't change the external API
- *Simple to understand*: The flow is straightforward and predictable

The key insight is to emit the `input.KEYS.end` event immediately when the input promise resolves, rather than relying on proc-log's finally callback mechanism. This ensures that each input operation completes fully before the next one begins. Therefore, the prompt counter is not required anymore.

Cancelled user prompts are now handled directly in `catch` making `finally` obsolete.

As suggested, we're now passing the `silent` prompt parameter to the `inputHandler`. This way we can output a newline for password prompts, so they are not overwritten by subsequent output. 

### Summary

- Fixes race condition in input handler
- Avoids spinner interfering with input prompts
- Properly adds newlines with silent prompts or when input is cancelled by user
- Added tests to cover various prompt cases

😃

wraithgar commented on lib/utils/display.js:
```diff
@@ -355,12 +361,29 @@ class Display {
+
+        // Silent inputs like password prompts require special handling
+        // to preserve output when users hit enter (see input.KEYS.end).
+        this.#silentPrompt = meta[META]?.silentPrompt || false
```
This makes me nervous.  The display class is global, and state should be for the entire npm process, not for a single event.  It will likely mean we need to pass this along as we are calling the input events.


It seems here we have a more subtle issue in that `proc-log` itself isn't aware of `META` in the input start/end lifecycle.    It's an oversight and one we likely need to fix there instead.  Then our input handler in the display layer can rely on having the `META` passed into it as originally requested.

TLDR: the real problem here is that we need to ensure that `META` is maintained and passed along through the entire input lifecycle so that `input.KEYS.end` has access to it as originally requested.

mbtools replied:
I removed the META logic. The input options are now passed from read to end event.

mbtools committed:
Merge branch 'npm:latest' into spinner

mbtools committed:
Remove META logic, pass input options

mbtools commented:
It's now straightforward and behave well in manual testing. Please give it a try

wraithgar commented:
Howdy @mbtools just wanna say this is NOT being ignored.  We're focusing on the OIDC PR (and some other housekeeping items) before looking at this one and some other community PRs.

Thanks for your patience here.

wraithgar commented:
> TLDR: the real problem here is that we need to ensure that META is maintained and passed along through the entire input lifecycle so that input.KEYS.end has access to it as originally requested.

I think this is still a root problem.  You have 100% "solved" the problem here which is always the hardest part.  This is going to require a PR to `proc-log` first, which we can then utilize here.  I'll start work on that now.

Aeolun commented:
@wraithgar If you are not going to have time to solve this problem any time soon, maybe it would make sense to just merge this PR until you can get around to the proper solution? It's gotten to the point where my LLM tells me NPM displaying a spinner during the login password prompt is 'expected behavior'.

edraft-123 commented:
@wraithgar Since it's still a bug, will it ever be merged?

mbtools committed:
Merge branch 'latest' into spinner

wraithgar commented:
https://github.com/npm/cli/pull/8779 updates proc-log in npm to have the code this PR needs to move forward

mbtools committed:
Merge branch 'npm:latest' into spinner

mbtools committed:
Replace process.emit with input methods

mbtools committed:
Fix options

mbtools commented:
All events now go via `proc-log` 🎉. I improved the email prompt a bit, too (see #8780).

mbtools committed:
Tweak args

wraithgar commented on lib/utils/display.js:
```diff
@@ -330,20 +330,38 @@ class Display {
+      case input.KEYS.end: {
         log.resume()
+        // For silent prompts (like password), add newline to preserve output
+        const [options] = args
```
the `withMeta` code has parsed the args at this point and looked for the meta symbol to pull out meta args and put them into `meta`.  You can see how we use it in the write path.  

At this point in the code it means is you can just optimistically ask for things like `meta.silent`.

wraithgar replied:
This will also mean adding META to the input events in read-user-info.js.

Examples of its use here in the cli can be found in `lib/utils/open-url.js`, `lib/commands/view.js` and `lib/commands/token.js`

mbtools replied:
Now I see how META works 😃  Thanks!

wraithgar replied:
It is NOT an intuitive thing, for sure.  Really glad this is getting near to completion!

mbtools committed:
Switch to META

mbtools committed:
Remove obsolete comment

wraithgar commented:
This is looking really good now.  As a reminder we are approaching a very long weekend here in the U.S. so an actual review (and, one may hope, a merge) won't likely happen till next week.

mbtools commented:
Enjoy a happy Thanksgiving🎉

mbtools commented:
Thank you for merging 🙏 and expanding the description. 

The issue wasn't so much the newlines but the fact that the spinner kept going while prompting the user for input. Now it shows a normal blinking cursor as one would expect 😉

wraithgar commented:
Ah, sorry I guess I missed the mark on the PR title a bit.  The changelog entry links here so those interested can dig way into things if they really want.
